src/dataflow-types/src/client/controller.rs

```diff
@@ -229,20 +256,53 @@ where
                         .await?;
                 let client = RemoteClient::new(&service.addresses("controller"));
                 let client: Box<dyn ComputeClient<T>> = Box::new(client);
-                self.compute_mut(instance)
+                self.compute_mut(instance_id)
                     .unwrap()
-                    .add_replica("default".into(), client)
+                    .add_replica(replica_name, client)
                     .await;
             }
         }
 
         Ok(())
     }
+
+    /// Removes a replica from an instance, including its service in the
+    /// orchestrator.
+    pub async fn drop_replica(
+        &mut self,
+        instance_id: ComputeInstanceId,
+        replica_id: ReplicaId,
+        config: ConcreteComputeInstanceReplicaConfig,
+    ) -> Result<(), anyhow::Error> {
+        let replica_name = generate_replica_service_name(instance_id, replica_id);
+        if let ConcreteComputeInstanceReplicaConfig::Managed {
+            size_config: _size_config,
+        } = config
+        {
+            let OrchestratorConfig { orchestrator, .. } = &self.orchestrator;
+            orchestrator
+                .namespace("compute")
+                .drop_service(&replica_name)
+                .await?;
+        }
+        let mut compute = self.compute_mut(instance_id).unwrap();
+        compute.remove_replica(&replica_name);
```

Comment:
We might want this to be the other way around: first remove the replica, then tell the orchestrator about dropping the service. We might race for a moment otherwise when the service terminates and we try to reconnect. Not sure if it can actually happen due to being effectively single-threaded, but who knows.